crates/ironrdp-cliprdr-format/src/bitmap.rs

```diff
@@ -0,0 +1,663 @@
+            profile_size,
+        })
+    }
+}
+
+fn validate_v1_header(header: &BitmapInfoHeader) -> Result<(), BitmapError> {
+    if header.width < 0 {
+        return Err(BitmapError::Unsupported("negative width"));
+    }
+
+    if header.width == 0 || header.height == 0 {
+        return Err(BitmapError::InvalidSize);
+    }
+
+    // In the modern world bitmaps with bpp < 24 are rare, and it is even more rare for the bitmaps
+    // which are placed on the clipboard as DIBs, therefore we could safely skip the support for
+    // such bitmaps.
+    const SUPPORTED_BIT_COUNT: &[u16] = &[24, 32];
+
+    if !SUPPORTED_BIT_COUNT.contains(&header.bit_count) {
+        return Err(BitmapError::Unsupported("unsupported bit count"));
+    }
+
+    // We support only uncompressed DIB bitmaps as it is the most common case for clipboard-copied
+    // bitmaps.
+    const SUPPORTED_COMPRESSION: &[BitmapCompression] = &[BitmapCompression::RGB];
+    if !SUPPORTED_COMPRESSION.contains(&header.compression) {
+        return Err(BitmapError::Unsupported("unsupported compression"));
+    }
+
+    // This is only relevant for bitmaps with bpp < 24, which are not supported.
+    if header.clr_used != 0 {
+        return Err(BitmapError::Unsupported("color table is not supported"));
+    }
+
+    Ok(())
+}
+
+fn validate_v5_header(header: &BitmapV5Header) -> Result<(), BitmapError> {
+    validate_v1_header(&header.header_v1)?;
+
+    const SUPPORTED_COLOR_SPACE: &[ColorSpace] = &[
+        ColorSpace::SRGB,
+        // Assume that Windows color space is sRGB, either way we don't have enough information on
+        // the clipboard to convert it to other color spaces.
+        ColorSpace::WINDOWS,
+    ];
+
+    if !SUPPORTED_COLOR_SPACE.contains(&header.color_space) {
+        return Err(BitmapError::Unsupported("not supported color space"));
+    }
+
+    Ok(())
+}
+
+struct PngEncoderInput {
+    frame_buffer: Vec<u8>,
+    width: usize,
+    height: usize,
+    color_type: png::ColorType,
+}
+
+/// From MS docs:
+/// For uncompressed RGB formats, the minimum stride is always the image width in bytes, rounded
+/// up to the nearest DWORD. You can use the following formula to calculate the stride and image
+/// size:
+/// ```
+/// stride = ((((biWidth * biBitCount) + 31) & ~31) >> 3);
+/// biSizeImage = abs(biHeight) * stride;
+/// ```
+fn bmp_stride(width: usize, bit_count: usize) -> usize {
+    (((width * bit_count) + 31) & !31) >> 3
+}
+
+fn transform_bitmap(
+    header: &BitmapInfoHeader,
+    input: &[u8],
+    preserve_alpha: bool,
+) -> Result<PngEncoderInput, BitmapError> {
+    // If height is positive, DIB is bottom-up, but target PNG format is top-down.
+    let flip = header.height > 0;
+
+    let width: usize = cast_int!(BitmapInfoHeader::NAME, "biWidth", header.width).expect("width is positive");
+    let height: usize = cast_int!(BitmapInfoHeader::NAME, "biHeight", header.height.abs()).expect("height is positive");
```

Comment:
Yep, it makes sense at least for height, negative height as per wingdi, requires to flip the vertical coordinates for the image (scanlines top-down instead of bottom-up). Negative width does not have a special meaning as per wingdi, but i think we could keep it i32 just for consistency

Reply:
Fair enough! :+1: